### PR TITLE
Support terminal resizing for exec/attach/run

### LINKED ISSUE #2284: "osc exec" results in a dumb terminal and no tty
It's important to the users coming from 2.x that `osc exec` be roughly the same experience as ssh in earlier versions.  This issue is meant to track progress towards making this a reality for GA.

#### Current State ####
Right now we're using nsenter under the covers.  As best I can tell nsenter doesn't actually create a tty nor have the ability to.  In addition a `dumb` terminal is set up by default.  This means:

* Using 'less' on a file will result in the following output:

          root@nginx:/var/log# less yum.log 
          WARNING: terminal is not fully functional
          yum.log  (press RETURN)

 This could be addressed today by setting a reasonable value for `$TERM` here: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools/manager.go#L994.  I hardcoded `TERM=xterm` in my environment and that mostly worked.  The trick with this approach is that it would be more correct to automatically set this based on the terminal the user is actually using.

 It's important to add that simply connecting with `osc exec` and setting `TERM=xterm` does not work correctly today in both Kubernetes and Origin.  Try doing so, running `less` and then using the up and down arrows and you'll see the problem.  Even if `Docker exec` doesn't set a default terminal in docker-1.6.0-6.el7.x86_64 it's doing something to make things work as expected.

* Without a TTY things like screen and tmux won't work.  I suspect we'll get complaints about that.  The workaround right now is to run `script /dev/null -c tmux` to trick the executable you want to run.  That really seems like a hack.

### PR DESCRIPTION
**Pull request by ncdc**
Add support for terminal resizing for exec, attach, and run. Note that for Docker, exec sessions
inherit the environment from the primary process, so if the container was created with tty=false,
that means the exec session's TERM variable will default to "dumb". Users can override this by
setting TERM=xterm (or whatever is appropriate) to get the correct "smart" terminal behavior.

Fixes #2284

### CODE REVIEW

**Comment by ncdc** [comment]
@deads2k please review the revert commit and the 2 carry commits that replace it.

@openshift/ui-review please review the "Support Windows terminal resizing" commit to make sure that I got every place we potentially run the CLI and pass in stdin/stdout/stderr - we have to use `term.StdStreams()` instead to get the right handle for stdout on Windows.

@bparees please review the last commit that affects binary build streaming.

**Comment by ncdc** [comment]
@smarterclayton @sttts fyi

**Comment by ncdc** [comment]
@csrwng can you try this on Windows?

File: pkg/build/registry/buildconfiginstantiate/rest.go [hunk 1]
```diff
@@ -219,7 +219,11 @@ func (h *binaryInstantiateHandler) handle(r io.Reader) (runtime.Object, error) {
 	if err != nil {
 		return nil, errors.NewInternalError(fmt.Errorf("unable to connect to server: %v", err))
 	}
-	if err := exec.Stream(kubeletremotecommand.SupportedStreamingProtocols, r, nil, nil, false); err != nil {
+	streamOptions := remotecommand.StreamOptions{
+		SupportedProtocols: kubeletremotecommand.SupportedStreamingProtocols,
+		Stdin:              r,
+	}
+	if err := exec.Stream(streamOptions); err != nil {
```
Comment (bparees) [inline pkg/build/registry/buildconfiginstantiate/rest.go]: lgtm

**Comment by deads2k** [comment]
reverts and carrys lgtm.

Since they exist to allow the latest client perform exec against 3.0.0 server that hasn't had any updates for almost a year, I don't see reason to fix these again.  Next time they need to be reverted, we can just drop them.

**Comment by csrwng** [comment]
@ncdc, I get the following with exec:
```
C:\Users\cewong\go\src\github.com\openshift\origin>oc exec simple-ruby-1-479pc -t -i -- bash
Unable to use a TTY - input is not a terminal or the right kind of file
```
(Same with powershell)

With `oc rsh` it doesn't give me a message, it just hangs there.

**Comment by csrwng** [comment]
The linux client works ok

**Comment by ncdc** [comment]
Investigating how to fix Windows

**Comment by ncdc** [comment]
For some reason using stdin from `term.StdStreams` on Windows breaks input with `oc login` - you see `Username:` and when you type, it doesn't echo. More fun stuff to dig into.

**Comment by ncdc** [comment]
So it turns out that what you get back in Windows from `term.StdStreams()` isn't compatible with things like `bufio.Reader` and `bufio.Scanner`. What Docker did where they require a prompt (`docker login`) is explicitly use `os.Stdin`.

I'm wondering now if maybe what @sttts had originally in his Windows work is what we should do - use `os.Stdin/out/err` everywhere, and only switch to using the values from `term.StdStreams()` for exec, run, and attach. What do you all think? @fabianofranz @smarterclayton @liggitt

**Comment by smarterclayton** [comment]
I would prefer to keep term.StdStreams isolated to only where it has to be.

**Comment by ncdc** [comment]
Yeah, that's my plan (and I'm going to do this in Kube so origin will get
it without any extra work).

**Comment by ncdc** [comment]
@bparees @smarterclayton @csrwng updated to pull in my fixes from upstream for term.StdStreams. Please review the last commit.

**Comment by ncdc** [comment]
[test]

**Comment by csrwng** [comment]
Last commit LGTM

**Comment by smarterclayton** [comment]
Looks ok to me to.  Anything else anyone wants to say?

**Comment by jwforres** [comment]
Just that I am anxiously waiting to hit the green button on this
https://github.com/kubernetes-ui/container-terminal/pull/7

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**